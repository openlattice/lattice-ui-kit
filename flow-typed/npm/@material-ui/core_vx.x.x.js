// flow-typed signature: 7ed29ef6da93a10ffdf64ee435c9faae
// flow-typed version: <<STUB>>/@material-ui/core_v4.5.2/flow_v0.117.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   '@material-ui/core'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the
 * community by sending a pull request to:
 * https://github.com/flowtype/flow-typed
 */

declare module '@material-ui/core' {
  declare module.exports: any;
}

declare module '@material-ui/core/styles' {
  declare module.exports: any;
}
