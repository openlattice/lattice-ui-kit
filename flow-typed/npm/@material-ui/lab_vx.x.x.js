// flow-typed signature: a13bdd1d57cdd924624da6dae303f597
// flow-typed version: <<STUB>>/@material-ui/lab_v4.0.0-alpha.40/flow_v0.117.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   '@material-ui/lab'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the
 * community by sending a pull request to:
 * https://github.com/flowtype/flow-typed
 */

declare module '@material-ui/lab' {
  declare module.exports: any;
}
