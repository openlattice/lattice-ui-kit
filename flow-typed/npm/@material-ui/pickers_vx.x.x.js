// flow-typed signature: 961daa4d7c37688be7a4da00b615a7ba
// flow-typed version: <<STUB>>/@material-ui/pickers_v~3.2.10/flow_v0.123.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   '@material-ui/pickers'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the
 * community by sending a pull request to:
 * https://github.com/flowtype/flow-typed
 */

declare module '@material-ui/pickers' {
  declare module.exports: any;
}
