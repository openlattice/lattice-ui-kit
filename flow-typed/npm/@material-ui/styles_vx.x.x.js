// flow-typed signature: 5f9d61d50037b50ea5904b8abd1b18ff
// flow-typed version: <<STUB>>/@material-ui/styles_v~4.9.0/flow_v0.123.0

/**
 * This is an autogenerated libdef stub for:
 *
 *   '@material-ui/styles'
 *
 * Fill this stub out by replacing all the `any` types.
 *
 * Once filled out, we encourage you to share your work with the
 * community by sending a pull request to:
 * https://github.com/flowtype/flow-typed
 */

declare module '@material-ui/styles' {
  declare module.exports: any;
}
