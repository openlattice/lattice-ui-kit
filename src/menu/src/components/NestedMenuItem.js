// @flow
import React, { useImperativeHandle, useRef, useState } from 'react';

import clsx from 'clsx';
import { faChevronRight } from '@fortawesome/pro-regular-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { ListItemSecondaryAction, Menu, MenuItem } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import type { MenuItemProps as MuiMenuItemProps } from '@material-ui/core';

import { NEUTRAL } from '../../../colors';
import { ARROW_LEFT, ARROW_RIGHT, ESC } from '../../../utils/keycodes';

const ChevronRight = () => <FontAwesomeIcon icon={faChevronRight} color={NEUTRAL.N500} size="sm" />;

// open NestedMenuItem maintains themed hover style
const useMenuItemStyles = makeStyles((theme) => ({
  root: (props) => ({
    backgroundColor: props.open ? theme.palette.action.hover : 'transparent'
  })
}));

const useNestedMenuItemStyles = makeStyles((theme) => ({
  root: {
    '&:focus': {
      outline: 'none',
      backgroundColor: theme.palette.action.selected
    }
  }
}));

/**
 * Use as a drop-in replacement for `<MenuItem>` when you need to add cascading
 * menu elements as children to this component.
 */
type Props = typeof MenuItem;

const NestedMenuItem = React.forwardRef<MuiMenuItemProps, MenuItem>((props :Props, ref) => {
  const {
    ContainerProps: ContainerPropsProp = {},
    children,
    className,
    elevation,
    label,
    parentMenuOpen,
    rightIcon = <ChevronRight />,
    tabIndex: tabIndexProp,
    ...MenuItemProps
  } = props;

  const { ref: containerRefProp, ...ContainerProps } = ContainerPropsProp;

  const menuItemRef = useRef(null);
  useImperativeHandle(ref, () => menuItemRef.current);

  const containerRef = useRef(null);
  useImperativeHandle(containerRefProp, () => containerRef.current);

  const menuContainerRef = useRef(null);

  const [isSubMenuOpen, setIsSubMenuOpen] = useState(false);

  const handleMouseEnter = (event) => {
    setIsSubMenuOpen(true);

    if (ContainerProps?.onMouseEnter) {
      ContainerProps.onMouseEnter(event);
    }
  };
  const handleMouseLeave = (event) => {
    setIsSubMenuOpen(false);

    if (ContainerProps?.onMouseLeave) {
      ContainerProps.onMouseLeave(event);
    }
  };

  // Check if any immediate children are active
  const isSubMenuFocused = () => {
    const active = containerRef.current?.ownerDocument?.activeElement;
    const nestedChildren = menuContainerRef.current?.children ?? [];
    for (let i = 0; i < nestedChildren.length; i += 1) {
      if (nestedChildren[i] === active) {
        return true;
      }
    }
    return false;
  };

  const handleKeyDown = (event) => {
    if (event.key === ESC) {
      setIsSubMenuOpen(false);
      return;
    }

    if (isSubMenuFocused()) {
      event.stopPropagation();
    }

    const active = containerRef.current?.ownerDocument?.activeElement;

    if (event.key === ARROW_LEFT && isSubMenuFocused()) {
      setIsSubMenuOpen(false);
      containerRef.current?.focus();
    }

    if (
      event.key === ARROW_RIGHT
      && event.target === containerRef.current
      && event.target === active
    ) {
      setIsSubMenuOpen(true);
      const firstChild = menuContainerRef.current?.children[0];
      firstChild?.focus();
    }
  };

  const open = !!(isSubMenuOpen && parentMenuOpen);
  const menuItemClasses = useMenuItemStyles({ open });
  const nestedMenuItemClasses = useNestedMenuItemStyles();

  // Root element must have a `tabIndex` attribute for keyboard navigation
  let tabIndex;
  if (!props.disabled) {
    tabIndex = tabIndexProp !== undefined ? tabIndexProp : -1;
  }

  return (
    <div
        className={clsx(nestedMenuItemClasses.root)}
        /* eslint-disable-next-line react/jsx-props-no-spreading */
        {...ContainerProps}
        onKeyDown={handleKeyDown}
        onMouseEnter={handleMouseEnter}
        onMouseLeave={handleMouseLeave}
        ref={containerRef}
        role="menu"
        tabIndex={tabIndex}>
      <MenuItem
          className={clsx(menuItemClasses.root, className)}
          ref={menuItemRef}
          /* eslint-disable-next-line react/jsx-props-no-spreading */
          {...MenuItemProps}>
        {label}
        <ListItemSecondaryAction>{rightIcon}</ListItemSecondaryAction>
      </MenuItem>
      <Menu
        // Set pointer events to 'none' to prevent the invisible Popover div
        // from capturing events for clicks and hovers
          anchorEl={menuItemRef.current}
          anchorOrigin={{
            vertical: 'top',
            horizontal: 'right'
          }}
          autoFocus={false}
          disableAutoFocus
          disableEnforceFocus
          elevation={elevation}
          getContentAnchorEl={null}
          onClose={() => {
            setIsSubMenuOpen(false);
          }}
          onEntering={() => menuContainerRef.current?.children[0]?.focus()}
          open={open}
          style={{ pointerEvents: 'none' }}
          transformOrigin={{
            vertical: 'top',
            horizontal: 'left'
          }}>
        <div ref={menuContainerRef} style={{ pointerEvents: 'auto' }}>
          {children}
        </div>
      </Menu>
    </div>
  );
});

export default NestedMenuItem;
