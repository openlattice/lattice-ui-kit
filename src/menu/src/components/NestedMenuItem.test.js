import toJson from 'enzyme-to-json';
import { Menu, MenuItem } from '@material-ui/core';
import { mount } from 'enzyme';

import NestedMenuItem from './NestedMenuItem';

const options = ['Option 1'];

describe('NestedMenuItem', () => {
  const defaultProps = {
    open: false,
    anchorEl: () => document.createElement('div'),
  };

  test('should match snapshot', () => {
    const wrapper = mount(
      /* eslint-disable-next-line react/jsx-props-no-spreading */
      <Menu {...defaultProps}>
        <NestedMenuItem label="Option 1">
          {options.map((option) => (
            <MenuItem
                key={`nested-${option}`}>
              {`Nested ${option}`}
            </MenuItem>
          ))}
        </NestedMenuItem>
      </Menu>
    );

    wrapper.setProps({
      open: true
    });

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe('event callbacks', () => {
    describe('enter', () => {
      test('should fire callback', () => {
        const mockHandleEnter = jest.fn();
        const wrapper = mount(
          /* eslint-disable-next-line react/jsx-props-no-spreading */
          <Menu {...defaultProps}>
            <NestedMenuItem
                ContainerProps={{
                  onMouseEnter: mockHandleEnter
                }}
                label="Option 1">
              {options.map((option) => (
                <MenuItem
                    key={`nested-${option}`}>
                  {`Nested ${option}`}
                </MenuItem>
              ))}
            </NestedMenuItem>
          </Menu>
        );

        wrapper.setProps({
          open: true
        });

        wrapper.update();

        const nestedMenuItem = wrapper.find(NestedMenuItem);

        expect(mockHandleEnter).toHaveBeenCalledTimes(0);
        nestedMenuItem.simulate('mouseenter');
        expect(mockHandleEnter).toHaveBeenCalledTimes(1);
      });
    });

    describe('exit', () => {
      test('should fire callback', () => {
        const mockHandleLeave = jest.fn();
        const wrapper = mount(
          /* eslint-disable-next-line react/jsx-props-no-spreading */
          <Menu {...defaultProps}>
            <NestedMenuItem
                ContainerProps={{
                  onMouseLeave: mockHandleLeave
                }}
                label="Option 1">
              {options.map((option) => (
                <MenuItem
                    key={`nested-${option}`}>
                  {`Nested ${option}`}
                </MenuItem>
              ))}
            </NestedMenuItem>
          </Menu>
        );

        wrapper.setProps({
          open: true
        });

        wrapper.update();

        const nestedMenuItem = wrapper.find(NestedMenuItem);

        expect(mockHandleLeave).toHaveBeenCalledTimes(0);
        nestedMenuItem.simulate('mouseleave');
        expect(mockHandleLeave).toHaveBeenCalledTimes(1);
      });
    });

    describe('focus', () => {
      test('should fire callback', () => {
        const mockOnFocus = jest.fn();
        const wrapper = mount(
          /* eslint-disable-next-line react/jsx-props-no-spreading */
          <Menu {...defaultProps}>
            <NestedMenuItem
                ContainerProps={{
                  onFocus: mockOnFocus
                }}
                label="Option 1">
              {options.map((option) => (
                <MenuItem
                    key={`nested-${option}`}>
                  {`Nested ${option}`}
                </MenuItem>
              ))}
            </NestedMenuItem>
          </Menu>
        );

        wrapper.setProps({
          open: true
        });

        wrapper.update();

        const nestedMenuItem = wrapper.find(NestedMenuItem);

        expect(mockOnFocus).toHaveBeenCalledTimes(0);
        nestedMenuItem.simulate('focus');
        expect(mockOnFocus).toHaveBeenCalledTimes(1);
      });
    });
  });
});
