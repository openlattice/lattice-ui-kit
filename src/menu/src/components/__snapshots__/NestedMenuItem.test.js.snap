// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NestedMenuItem should match snapshot 1`] = `
<WithStyles(ForwardRef(Menu))
  anchorEl={[Function]}
  open={true}
>
  <ForwardRef(Menu)
    anchorEl={[Function]}
    classes={
      Object {
        "list": "MuiMenu-list",
        "paper": "MuiMenu-paper",
      }
    }
    open={true}
  >
    <WithStyles(ForwardRef(Popover))
      PaperProps={
        Object {
          "classes": Object {
            "root": "MuiMenu-paper",
          },
        }
      }
      anchorEl={[Function]}
      anchorOrigin={
        Object {
          "horizontal": "left",
          "vertical": "top",
        }
      }
      getContentAnchorEl={[Function]}
      onEntering={[Function]}
      open={true}
      transformOrigin={
        Object {
          "horizontal": "left",
          "vertical": "top",
        }
      }
      transitionDuration="auto"
    >
      <ForwardRef(Popover)
        PaperProps={
          Object {
            "classes": Object {
              "root": "MuiMenu-paper",
            },
          }
        }
        anchorEl={[Function]}
        anchorOrigin={
          Object {
            "horizontal": "left",
            "vertical": "top",
          }
        }
        classes={
          Object {
            "paper": "MuiPopover-paper",
            "root": "MuiPopover-root",
          }
        }
        getContentAnchorEl={[Function]}
        onEntering={[Function]}
        open={true}
        transformOrigin={
          Object {
            "horizontal": "left",
            "vertical": "top",
          }
        }
        transitionDuration="auto"
      >
        <ForwardRef(Modal)
          BackdropProps={
            Object {
              "invisible": true,
            }
          }
          className="MuiPopover-root"
          container={
            <body
              style="padding-right: 0px; overflow: hidden;"
            >
              <div
                class="MuiPopover-root"
                role="presentation"
                style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
              >
                <div
                  aria-hidden="true"
                  style="z-index: -1; position: fixed; right: 0px; bottom: 0px; top: 0px; left: 0px; background-color: transparent;"
                />
                <div
                  data-test="sentinelStart"
                  tabindex="0"
                />
                <div
                  class="MuiPaper-root MuiMenu-paper MuiPopover-paper MuiPaper-elevation8 MuiPaper-rounded"
                  style="opacity: 1; transform: scale(1, 1); transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
                  tabindex="-1"
                >
                  <ul
                    class="MuiList-root MuiMenu-list MuiList-padding"
                    role="menu"
                    tabindex="-1"
                  >
                    <div
                      class="makeStyles-root-3"
                      role="menu"
                      tabindex="0"
                    >
                      <li
                        class="MuiListItem-container"
                      >
                        <div
                          aria-disabled="false"
                          class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root makeStyles-root-1 makeStyles-root-2 MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button MuiListItem-secondaryAction"
                          role="menuitem"
                          tabindex="-1"
                        >
                          Option 1
                          <span
                            class="MuiTouchRipple-root"
                          />
                        </div>
                        <div
                          class="MuiListItemSecondaryAction-root"
                        >
                          <svg
                            aria-hidden="true"
                            class="svg-inline--fa fa-chevron-right fa-w-8 fa-sm "
                            color="#8b8b90"
                            data-icon="chevron-right"
                            data-prefix="far"
                            focusable="false"
                            role="img"
                            viewBox="0 0 256 512"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M24.707 38.101L4.908 57.899c-4.686 4.686-4.686 12.284 0 16.971L185.607 256 4.908 437.13c-4.686 4.686-4.686 12.284 0 16.971L24.707 473.9c4.686 4.686 12.284 4.686 16.971 0l209.414-209.414c4.686-4.686 4.686-12.284 0-16.971L41.678 38.101c-4.687-4.687-12.285-4.687-16.971 0z"
                              fill="currentColor"
                            />
                          </svg>
                        </div>
                      </li>
                    </div>
                  </ul>
                </div>
                <div
                  data-test="sentinelEnd"
                  tabindex="0"
                />
              </div>
            </body>
          }
          open={true}
        >
          <ForwardRef(Portal)
            container={
              <body
                style="padding-right: 0px; overflow: hidden;"
              >
                <div
                  class="MuiPopover-root"
                  role="presentation"
                  style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
                >
                  <div
                    aria-hidden="true"
                    style="z-index: -1; position: fixed; right: 0px; bottom: 0px; top: 0px; left: 0px; background-color: transparent;"
                  />
                  <div
                    data-test="sentinelStart"
                    tabindex="0"
                  />
                  <div
                    class="MuiPaper-root MuiMenu-paper MuiPopover-paper MuiPaper-elevation8 MuiPaper-rounded"
                    style="opacity: 1; transform: scale(1, 1); transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
                    tabindex="-1"
                  >
                    <ul
                      class="MuiList-root MuiMenu-list MuiList-padding"
                      role="menu"
                      tabindex="-1"
                    >
                      <div
                        class="makeStyles-root-3"
                        role="menu"
                        tabindex="0"
                      >
                        <li
                          class="MuiListItem-container"
                        >
                          <div
                            aria-disabled="false"
                            class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root makeStyles-root-1 makeStyles-root-2 MuiMenuItem-gutters MuiListItem-gutters MuiListItem-button MuiListItem-secondaryAction"
                            role="menuitem"
                            tabindex="-1"
                          >
                            Option 1
                            <span
                              class="MuiTouchRipple-root"
                            />
                          </div>
                          <div
                            class="MuiListItemSecondaryAction-root"
                          >
                            <svg
                              aria-hidden="true"
                              class="svg-inline--fa fa-chevron-right fa-w-8 fa-sm "
                              color="#8b8b90"
                              data-icon="chevron-right"
                              data-prefix="far"
                              focusable="false"
                              role="img"
                              viewBox="0 0 256 512"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M24.707 38.101L4.908 57.899c-4.686 4.686-4.686 12.284 0 16.971L185.607 256 4.908 437.13c-4.686 4.686-4.686 12.284 0 16.971L24.707 473.9c4.686 4.686 12.284 4.686 16.971 0l209.414-209.414c4.686-4.686 4.686-12.284 0-16.971L41.678 38.101c-4.687-4.687-12.285-4.687-16.971 0z"
                                fill="currentColor"
                              />
                            </svg>
                          </div>
                        </li>
                      </div>
                    </ul>
                  </div>
                  <div
                    data-test="sentinelEnd"
                    tabindex="0"
                  />
                </div>
              </body>
            }
            disablePortal={false}
          />
        </ForwardRef(Modal)>
      </ForwardRef(Popover)>
    </WithStyles(ForwardRef(Popover))>
  </ForwardRef(Menu)>
</WithStyles(ForwardRef(Menu))>
`;
